forc-util/src/lib.rs

```diff
@@ -716,30 +716,40 @@ fn construct_window<'a>(
     let total_lines_of_highlight = end.line - start.line;
     debug_assert!(total_lines_in_input >= total_lines_of_highlight);
 
+    // these two are 0-based
     let mut current_line = 0;
     let mut lines_to_start_of_snippet = 0;
+
     let mut calculated_start_ix = None;
     let mut calculated_end_ix = None;
     let mut pos = 0;
+
```

Comment:
Since you initially `- 1` and then `+ 1` in multiple places, would it be simpler to leave the original variables and then just change the comparators i.e:

    - if current_line + NUM_LINES_BUFFER >= start.line && calculated_start_ix.is_none() {
    + if current_line + NUM_LINES_BUFFER > start.line && calculated_start_ix.is_none() {

and 

    - if current_line >= end.line + NUM_LINES_BUFFER && calculated_end_ix.is_none() {
    + if current_line > end.line + NUM_LINES_BUFFER && calculated_end_ix.is_none() {

Reply:
I simplified doing the inverse: starting the counter with 1, instead of zero.

Reply:
I don't think starting at `1` is correct - as a contrived example if `input` is an empty string, at the end of the function `master` currently does:

    start.line = 0;

Whereas your PR will be:

    start.line = 1;

As `start` is a reference to the function arguments, this will be used external to the call.

Reply:
Which is correct, given that `start` and `end` use 1-based indexing.
An empty code would call

```
construct_window(
        &mut LineCol { line: 1, col: 1 },
        LineCol { line: 1, col: 1 },
        &mut start,
        &mut end,
        "",
    )
```

And this should return 

```
[forc-util/src/lib.rs:772:5] start = 0
[forc-util/src/lib.rs:772:5] end = 0
[forc-util/src/lib.rs:772:5] s = LineCol {
    line: 1,
    col: 1,
}
[forc-util/src/lib.rs:772:5] r = ""
```

I think you meant the inverse, right? I am including more tests, nonetheless.

Reply:
In this case, I think master is incorrect. `start.line` uses 1-based index, and should never be zero. There is certainly a guard somewhere making `line = 0` work somehow.